react-cart-final.md

```diff
@@ -0,0 +1,1132 @@
+    it 'responds with success message' do
+      expect(JSON.parse(response.body)['message']).to eq 'Your order will be ready in 30 minutes!'
+		end
+
+		it 'sets the order attribute "finalized" to true' do
+			expect(order.reload.finalized).to eq true
+		end
+  end
+end
+```
+
+Run the specs you should and take notice of the error messages. 
+
+We need to add a new column in our orders database table. 
+
+`rails g migration AddFinalizedToOrders finalized:boolean`
+
+Before you migrate make sure that it is adding a column to the orders db table and make sure that the default value is set to false.
+
+It should look like this:
+
+
+```rb
+class AddFinalizedToOrders < ActiveRecord::Migration[6.0]
+  def change
+    add_column :orders, :finalized, :boolean, default: false
+  end
+end
+```
+run the migrations and make sure that the changes have been added to your schema file.
+
+The final step we need to take is to refactor our update action. 
+
+`app/controllers/api/orders_controller.rb`
+
+```rb
+	def update
+		order = Order.find(params[:id])
+		if params[:activity]
+			order.update_attribute(:finalized, true)
+      render json: { message: 'Your order will be ready in 30 minutes!' }
+    else
+      product = Product.find(params[:product_id])
+      order.order_items.create(product: product)
+      render json: create_json_response(order)
+    end
+  end
+```
+
+Remember that need to update our order serializer
+```rb
+attributes :id, :products_1, :products_2, :products_3, :products, :total, :order_total, :finalized
+```
+
+
+Run your spec they should be green by now. 
+
+Lets head back over to the client.
+
+## Finalizing and back to the client
+
+As always let's start with a new spec. 
+
+Add the following code to your feature test:
+
+```js
+it('user can finalize the order', () => {
+		cy.get('#product-2').within(() => {
+			cy.get('button').contains('Add to order').click()
+		})
+		cy.get('#product-3').within(() => {
+			cy.get('button').contains('Add to order').click()
+		})
+		cy.get('button').contains('View order').click()
+		cy.route({
+			method: 'PUT',
+			url: 'http://localhost:3000/api/orders/1',
+			response: { message: 'Your order will be ready in 30 minutes!' }
+		})
+		cy.get('button').contains('Confirm!').click()
+		cy.get('.message').should('contain', "Your order will be ready in 30 minutes!")
+}); 
+
+```
+
+We also want to make sure that we have som econfigration to use in the cypress tests. We are adding a delay depending on the different functions that cypress runs. 
+
+```js
+
+// cypress/support/commands.js
+const COMMAND_DELAY = 150;
+
+
+for (const command of ['get', 'visit', 'click', 'trigger', 'type', 'clear', 'reload', 'contains']) {
+    Cypress.Commands.overwrite(command, (originalFn, ...args) => {
+        const origVal = originalFn(...args);
+
+        return new Promise((resolve) => {
+            setTimeout(() => {
+                resolve(origVal);
+            }, COMMAND_DELAY);
+        });
+    });
+}
+```
```

Comment:
This only so that we can slow down the tests to see what is actually happening. I am not sure if this is relevant to the topic.